Linking Error: ../libATen.so.1: undefined reference to functions from NNPACK and some other libs...

Hi, when I build the CPU version from source, I met the following issue. I am using Python 3.5, Ubuntu 16.04. Does anyone know how to fix it? Thanks a lot!

> [ 89%] Linking CXX executable atest
> ../libATen.so.1: undefined reference to `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> src/ATen/test/CMakeFiles/atest.dir/build.make:96: recipe for target 'src/ATen/test/atest' failed
> make[2]: *** [src/ATen/test/atest] Error 1
> CMakeFiles/Makefile2:243: recipe for target 'src/ATen/test/CMakeFiles/atest.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/atest.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> [ 90%] Linking CXX executable broadcast_test
> ../libATen.so.1: undefined reference to `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> src/ATen/test/CMakeFiles/broadcast_test.dir/build.make:96: recipe for target 'src/ATen/test/broadcast_test' failed
> make[2]: *** [src/ATen/test/broadcast_test] Error 1
> CMakeFiles/Makefile2:354: recipe for target 'src/ATen/test/CMakeFiles/broadcast_test.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/broadcast_test.dir/all] Error 2
> [ 90%] Linking CXX executable basic
> ../libATen.so.1: undefined reference to `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> src/ATen/test/CMakeFiles/basic.dir/build.make:96: recipe for target 'src/ATen/test/basic' failed
> make[2]: *** [src/ATen/test/basic] Error 1
> CMakeFiles/Makefile2:317: recipe for target 'src/ATen/test/CMakeFiles/basic.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/basic.dir/all] Error 2
> [ 91%] Linking CXX executable native_test
> ../libATen.so.1: undefined reference to `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> src/ATen/test/CMakeFiles/native_test.dir/build.make:96: recipe for target 'src/ATen/test/native_test' failed
> make[2]: *** [src/ATen/test/native_test] Error 1
> CMakeFiles/Makefile2:280: recipe for target 'src/ATen/test/CMakeFiles/native_test.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/native_test.dir/all] Error 2
> Makefile:127: recipe for target 'all' failed
> make: *** [all] Error 2

Hi @apaszke Thanks for your response! The full build log is as follows. It seems that NNPACK had been installed. ` Found NNPACK    (include: /usr/local/include, library: /usr/local/lib/libnnpack.so)`

> running install
> running build_deps
> CUDA_TOOLKIT_ROOT_DIR not found or specified
> -- Could NOT find CUDA (missing:  CUDA_TOOLKIT_ROOT_DIR CUDA_NVCC_EXECUTABLE CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) (Required is at least version "5.5")
> -- Found gcc >=5 and CUDA <= 7.5, adding workaround C++ flags
> -- Automatic GPU detection failed. Building for common architectures.
> -- Autodetected CUDA architecture(s): 3.0;3.5;5.0;5.2+PTX
> -- Could not find CUDA with FP16 support, compiling without torch.CudaHalfTensor
> -- Removing -DNDEBUG from compile flags
> -- Compiling with OpenMP support
> -- MAGMA not found. Compiling without MAGMA support
> -- Could not find hardware support for NEON on this machine.
> -- No OMAP3 processor on this machine.
> -- No OMAP4 processor on this machine.
> -- SSE2 Found
> -- SSE3 Found
> -- AVX Found
> -- AVX2 Found
> -- Atomics: using GCC intrinsics
> -- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_sequential - mkl_core - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - gomp - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - iomp5 - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_gf_lp64 - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf_lp64 - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf_lp64: not found
> -- Checking for [mkl_gf - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_gf - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_gf: not found
> -- Checking for [mkl_intel_lp64 - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel_lp64 - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel_lp64: not found
> -- Checking for [mkl_intel - mkl_gnu_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl_intel - mkl_intel_thread - mkl_core - pthread - m - dl]
> --   Library mkl_intel: not found
> -- Checking for [mkl - guide - pthread - m]
> --   Library mkl: not found
> -- MKL library not found
> -- Checking for [openblas]
> --   Library openblas: /usr/lib/libopenblas.so
> -- Found a library with BLAS API (open).
> -- Found a library with LAPACK API. (open)
> disabling CUDA because NO_CUDA is set
> -- Could NOT find CUDNN (missing:  CUDNN_INCLUDE_DIR CUDNN_LIBRARY) 
> -- CuDNN not found. Compiling without CuDNN support
> -- Found NNPACK    (include: /usr/local/include, library: /usr/local/lib/libnnpack.so)
> -- Found PTHREADPOOL (library: /usr/local/lib/libpthreadpool.so)
> -- Using python found in /usr/bin/python3
> disable contrib because ATEN_NO_CONTRIB is set
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /home/pytorch-git/torch/lib/build/aten
> [  1%] Built target aten_files_are_generated
> [ 84%] Built target ATen
> [ 85%] Linking CXX executable atest
> [ 85%] Linking CXX executable basic
> [ 86%] Linking CXX executable native_test
> [ 87%] Linking CXX executable broadcast_test
> .../libATen.so.1: undefined reference to .`nnp_convolution_kernel_gradient'
> ..//libATen.so.1: undefinedlibATen.so.1 reference to `nnp_convolution_input_gradient': 
> ../libATen.so.1undefined : undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference toreference ` nnp_initializeto'
>  collect2: error: ld returned 1 exit status
> `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../src/ATen/test/CMakeFiles/atest.dir/build.make:96: recipe for target 'src/ATen/test/atest' failed
> make[2]: *** [src/ATen/test/atest] Error 1
> libATen.so.1: undefined reference to `nnp_convolution_inference'
> CMakeFiles/Makefile2:243: recipe for target 'src/ATen/test/CMakeFiles/atest.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/atest.dir/all] Error 2
> .make[1]: *** Waiting for unfinished jobs....
> ./libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> src/ATen/test/CMakeFiles/broadcast_test.dir/build.make:96: recipe for target 'src/ATen/test/broadcast_test' failed
> make[2]: *** [src/ATen/test/broadcast_test] Error 1
> CMakeFiles/Makefile2:354: recipe for target 'src/ATen/test/CMakeFiles/broadcast_test.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/broadcast_test.dir/all] Error 2
> ../libATen.so.1: undefined reference to `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> ../libATen.so.1: undefined reference to `nnp_convolution_kernel_gradient'
> ../libATen.so.1: undefined reference to `nnp_convolution_input_gradient'
> ../libATen.so.1: undefined reference to `pthreadpool_create'
> ../libATen.so.1: undefined reference to `nnp_convolution_output'
> ../libATen.so.1: undefined reference to `nnp_convolution_inference'
> ../libATen.so.1: undefined reference to `nnp_initialize'
> collect2: error: ld returned 1 exit status
> src/ATen/test/CMakeFiles/native_test.dir/build.make:96: recipe for target 'src/ATen/test/native_test' failed
> make[2]: *** [src/ATen/test/native_test] Error 1
> CMakeFiles/Makefile2:280: recipe for target 'src/ATen/test/CMakeFiles/native_test.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/native_test.dir/all] Error 2
> src/ATen/test/CMakeFiles/basic.dir/build.make:96: recipe for target 'src/ATen/test/basic' failed
> make[2]: *** [src/ATen/test/basic] Error 1
> CMakeFiles/Makefile2:317: recipe for target 'src/ATen/test/CMakeFiles/basic.dir/all' failed
> make[1]: *** [src/ATen/test/CMakeFiles/basic.dir/all] Error 2
> Makefile:127: recipe for target 'all' failed
> make: *** [all] Error 2

--- a/aten/cmake/FindNNPACK.cmake
+++ b/aten/cmake/FindNNPACK.cmake
@@ -12,6 +12,7 @@
 #  NNPACK_LIBRARYRARY_DIRS
 
 include(FindPackageHandleStandardArgs)
+include(CheckSymbolExists)
 
 set(NNPACK_ROOT_DIR "" CACHE PATH "Folder contains NNPACK")
 
@@ -26,17 +27,31 @@ find_library(NNPACK_LIBRARY nnpack
     PATHS ${NNPACK_ROOT_DIR}
     PATH_SUFFIXES lib lib64)
 
+find_library(CPUINFO_LIBRARY cpuinfo
+    PATHS ${NNPACK_ROOT_DIR}
+    PATH_SUFFIXES lib lib64)
+
 find_library(PTHREADPOOL_LIBRARY pthreadpool
     PATHS ${NNPACK_ROOT_DIR}
     PATH_SUFFIXES lib lib64)
 
-find_package_handle_standard_args(NNPACK DEFAULT_MSG NNPACK_INCLUDE_DIR NNPACK_LIBRARY PTHREADPOOL_LIBRARY)
+find_package_handle_standard_args(NNPACK DEFAULT_MSG NNPACK_INCLUDE_DIR NNPACK_LIBRARY CPUINFO_LIBRARY PTHREADPOOL_LIBRARY)
 
 if(NNPACK_FOUND)
   set(NNPACK_INCLUDE_DIRS ${NNPACK_INCLUDE_DIR})
-  set(NNPACK_LIBRARIES ${NNPACK_LIBRARY} ${PTHREADPOOL_LIBRARY})
-  message(STATUS "Found NNPACK    (include: ${NNPACK_INCLUDE_DIR}, library: ${NNPACK_LIBRARY})")
-  message(STATUS "Found PTHREADPOOL (library: ${PTHREADPOOL_LIBRARY})")
-  mark_as_advanced(NNPACK_ROOT_DIR NNPACK_LIBRARY_RELEASE NNPACK_LIBRARY_DEBUG
-                                 NNPACK_LIBRARY NNPACK_INCLUDE_DIR)
+  set(NNPACK_LIBRARIES ${NNPACK_LIBRARY} ${CPUINFO_LIBRARY} ${PTHREADPOOL_LIBRARY})
+
+  list(APPEND CMAKE_REQUIRED_LIBRARIES ${NNPACK_LIBRARIES})
+  list(APPEND CMAKE_REQUIRED_INCLUDES ${NNPACK_INCLUDE_DIRS})
+  check_symbol_exists(nnp_convolution_kernel_gradient "nnpack.h" NNPACK_HAS_INFERENCE)
+
+  if(NNPACK_HAS_INFERENCE)
+    message(STATUS "Found NNPACK      (include: ${NNPACK_INCLUDE_DIR}, library: ${NNPACK_LIBRARY})")
+    message(STATUS "Found CPUINFO     (library: ${CPUINFO_LIBRARY})")
+    message(STATUS "Found PTHREADPOOL (library: ${PTHREADPOOL_LIBRARY})")
+    mark_as_advanced(NNPACK_ROOT_DIR NNPACK_LIBRARY_RELEASE NNPACK_LIBRARY_DEBUG
+                                   NNPACK_LIBRARY NNPACK_INCLUDE_DIR)
+  else()
+    message(STATUS "Refusing to use incomplete NNPACK (include: ${NNPACK_INCLUDE_DIR}, library: ${NNPACK_LIBRARY}); try reinstalling NNPACK without --inference-only")
+  endif()
 endif()

